src/jit/codegenarm64.cpp

```diff
@@ -4578,49 +4578,139 @@ void CodeGen::genSIMDIntrinsicGetItem(GenTreeSIMD* simdNode)
     // - the source of SIMD type (op1)
     // - the index of the value to be returned.
     genConsumeOperands(simdNode);
-    regNumber srcReg = op1->gtRegNum;
 
-    // TODO-ARM64-CQ Optimize SIMDIntrinsicGetItem
-    // Optimize the case of op1 is in memory and trying to access ith element.
-    assert(op1->isUsedFromReg());
+    emitAttr baseTypeSize  = emitTypeSize(baseType);
+    unsigned baseTypeScale = genLog2(EA_SIZE_IN_BYTES(baseTypeSize));
 
-    emitAttr baseTypeSize = emitTypeSize(baseType);
-
-    if (op2->IsCnsIntOrI())
+    if (op1->isContained())
     {
-        assert(op2->isContained());
+        // Optimize the case of op1 is in memory and trying to access ith element.
+        assert(!op1->isUsedFromReg());
+
+        if (op1->OperIsLocal())
+        {
+            unsigned varNum = op1->gtLclVarCommon.gtLclNum;
+
+            if (op2->IsCnsIntOrI())
+            {
+                assert(op2->isContained());
 
-        ssize_t index = op2->gtIntCon.gtIconVal;
+                ssize_t index = op2->gtIntCon.gtIconVal;
 
-        if (getEmitter()->isValidVectorIndex(emitTypeSize(simdType), baseTypeSize, index))
+                if (getEmitter()->isValidVectorIndex(emitTypeSize(simdType), baseTypeSize, index))
+                {
+                    // Only generate code for the get if the index is valid
+                    // Otherwise generated code will throw
+                    int offset = (int)index * genTypeSize(baseType);
+
+                    instruction ins = ins_Load(baseType);
+                    baseTypeSize    = varTypeIsFloating(baseType)
+                                       ? baseTypeSize
+                                       : getEmitter()->emitInsAdjustLoadStoreAttr(ins, baseTypeSize);
+
+                    getEmitter()->emitIns_R_S(ins, baseTypeSize, targetReg, varNum, offset);
+                }
+            }
+            else
+            {
+                assert(!op2->isContained());
+
+                regNumber indexReg = op2->gtRegNum;
+                regNumber tmpReg   = simdNode->ExtractTempReg();
+
+                assert(genIsValidIntReg(indexReg));
+                assert(genIsValidIntReg(tmpReg));
+
+                // Load the address of varNum
+                getEmitter()->emitIns_R_S(INS_lea, EA_PTRSIZE, tmpReg, varNum, 0);
```

Comment:
It seems that it might be cleaner if you first generate code to compute the "baseReg", which in this case would be `tmpReg`, and in the non-contained case would be the address of `op1->AsIndir()->Addr()`. Then the other code (e.g. the handling of a constant `op2`) could be shared.

Reply:
I am not sure precisely what you are saying.  

I started with the approach used in xarch, but it was problematic due in part to ARM64's limited memory addressing modes.  

| case | op1 | op2 | current code |
| --- | --- | --- | --- | 
| 1(1.1.1) | contained local | const  | `emitIns_R_S(ins_Load(baseType), ..., offset)` |
| 2(1.1.2) | contained local | var  | `emitIns_R_S(INS_lea); emitIns_R_R_R_Ext(ins_Load(baseType),...baseReg, indexReg)` |
| 3(1.2.1) | contained indir | const  | `emitIns_R_R_I(ins_Load(baseType),..., baseReg, offset)` |
| 4(1.2.2) | contained indir | var  | `emitIns_R_R_R_Ext(ins_Load(baseType),..., baseReg, indexReg)` |
| 5(2.1) | reg | const  | `emitIns_R_R_I(INS_mov, ..., srcReg, index)` |
| 6(2.2) | reg | var  | `emitIns_R_S(INS_lea, ..., baseReg, simdInitTempVarNum);emitIns_R_R(INS_str, srcReg, baseReg); emitIns_R_R_R_Ext(ins_Load(baseType),..., baseReg, indexReg)` |

So `1(1.1.1)`, `3(1.2.1)`, `5(2.1)` want to remains separate.  Each typically emits a single instruction.

`2(1.1.2)`, `4(1.2.2)`, and `6(2.2)` could share the code generator for the last instruction.  So if I rearrange the conditions, it should simplify the code somewhat.

@CarolEidt  Is this what you had in mind?

Reply:
> So if I rearrange the conditions, it should simplify the code somewhat.

Yes, and if you restructured it so that you checked for the constant cases together, you could just have one place that checks `if (getEmitter()->isValidVectorIndex(emitTypeSize(simdType), baseTypeSize, index))`